Fix out-of-date map check (#1522)

ssoloff opened the pull request:
This PR fixes the broken out-of-date map check reported in issue #1522.

The out-of-date map check was using the wrong file name when looking for a map's download properties.  It only considered the property file `<mapName>.zip.properties`, which does not exist for map's downloaded from GitHub.  When the map download properties cannot be found, it is assumed the map is up-to-date.

Since migrating the maps to GitHub, `ResourceLoader` considers the following map zip file candidates when attempting to load a map:

1. `<mapName>.zip`
1. `<normalizedMapName>-master.zip`
1. `<normalizedMapName>.zip`

The fix for this issue was to modify the out-of-date map check to also consider all these map zip file names when attempting to locate the corresponding download properties file (the same name with `.properties` appended).

#### Functional changes
* The out-of-date map check in `MapDownloadController` now considers all possible map zip file candidates (see above) when looking for the corresponding download properties file.  Previously, it only considered candidate (1).

#### Functional issues for review
None.

#### Refactoring changes
* I extracted a new interface and several new methods in `MapDownloadController` in order to write unit tests for the out-of-date map check functionality.
* I modified `ResourceLoader#getPaths()` to use the new `ResourceLoader#getMapZipFileCandidates()` method in order to be DRY.

#### Refactoring issues for review
* The changes to `ResourceLoader#getPaths()` were not necessary to implement the fix in this PR.  As I said above, I simply did it to avoid duplication in map path candidates (which was essentially the root cause of this issue).  Please advise if that hunk should be reverted.
* The out-of-date map check has only ever considered maps whose shape is a zip file; maps with a directory shape are not supported by this feature.  Some of the names I used when refactoring reflect this limitation to make it explicitly clear.

#### Testing
I first confirmed I could reproduce the issue on `master` (1.9.0.0.4208).  I performed the following steps:

* Delete the `TRIPLEA_LAST_CHECK_FOR_MAP_UPDATES` key from the user preferences to force an out-of-date map check.
* Locally modify my `tutorial-master.zip.properties` and `the_pact_of_steel-master.zip.properties` files and change the `map.version` keys to `0.9` (they previously had the value `1.0`).  (I have a total of six maps installed, so there are still four maps present that are up-to-date.)
* Run TripleA.

After waiting for about a minute, I was not prompted to download the updated maps.

I then repeated the steps on this branch.  I confirmed I was prompted to download both updated maps.  I answered **Yes** to the prompt, downloaded the two maps in question, and exited TripleA.

I deleted the `TRIPLEA_LAST_CHECK_FOR_MAP_UPDATES` key once again and restarted TripleA.  I waited for about a minute to confirm I was not prompted to download any maps.

I also started a few games from maps whose shape is a zip file to confirm the changes in `ResourceLoader#getPaths()` did not break map loading.

ssoloff commented:
@simon33-2 

> Still leaves the weakness that you have to enter the download maps window to trigger the check though doesn't it?

No, the check occurs in the background once the TripleA main window is displayed--it does not require that you open the Download Maps window first (see `GameRunner#checkForUpdates()`).  However, depending on network latency, it can take several seconds for the notification to pop up.  During my testing, it took between 5-10 seconds before I was prompted.  (This is the same behavior as the tutorial map and out-of-date engine checks.)

The "weakness" that still exists is that if you answer Yes to the prompt to download the out-of-date maps, it simply opens the Download Maps window--it does not initiate the download(s) automatically.  I'll admit, I thought I introduced a regression when I was testing this because the prompt "Would you like to update (re-download) the following maps now?" implies the download(s) will start once I click Yes.  However, I confirmed that this was the behavior that existed before I made any changes (clicking Yes simply calls `DownloadMapsWindow#showDownloadMapsWindow()` without specifying a map name, which only opens the window without initiating any download(s)).

ssoloff commented:
Here's a demonstration of the prompt to update maps displayed without having to first open the Download Maps window:

![prompt-to-download-updated-maps](https://cloud.githubusercontent.com/assets/4826349/25879792/aa1bade0-3501-11e7-8ff0-d3a46f432909.png)

simon33-2 commented:
What is this looking at? It isn't working for me. I've updated the version in triplea_maps.yaml and it seems that it is looking somewhere else.

ssoloff commented:
> What is this looking at?

It gets the latest version from `triplea_maps.yaml` **in the repo.**  From `game_engine.properties`:

```properties
Map_List_File = http://raw.githubusercontent.com/triplea-game/triplea/master/triplea_maps.yaml?raw=true
```

So you can't just change your local `triplea_maps.yaml` to trigger an update (unless, of course, you also change the `Map_List_File` key in `game_engine.properties` to point to your local file).

The downloaded version is determined from the `map.version` key in the `<repoName>-<branchName>.zip.properties` file.  For example, from `~/triplea/downloadedMaps/tutorial-master.zip.properties` on my machine:

```properties
map.version=1.0
```

simon33-2 commented:
That's what I thought. I actually did those things but it doesn't seem to be working for me. Does it use the configured folder? I've changed it from ~/triplea/downloadedMaps.

I'm getting a couple of 302 errors so perhaps it doesn't work with a local triplea_maps.yaml.

ssoloff commented:
> Does it use the configured folder?

It uses whatever `ClientFileSystemHelper#getUserMapsFolder()` returns.  So, yes, it should pick up whatever map folder you've configured.

> I'm getting a couple of 302 errors so perhaps it doesn't work with a local triplea_maps.yaml.

These are actually expected.  I'm going to write up an issue to not log redirects as errors.

> but it doesn't seem to be working for me

The only other thing I can think of is that your most recent map update check was less than 30 days ago.  To force the code to perform a map update check, you need to delete the `TRIPLEA_LAST_CHECK_FOR_MAP_UPDATES` key in your user preferences.  If you are on *NIX, you'll probably find this key in the file `~/.java/.userPrefs/games/strategy/triplea/settings/prefs.xml`.  If you are on Windows, it is in the registry under `HKEY_CURRENT_USER\Software\JavaSoft\Prefs\games\strategy\triplea\settings`.

ssoloff commented:
@simon33-2, I did the following:

1. Changed the value of the `Map_List_File` key in `game_engine.properties`:

```properties
Map_List_File = triplea_maps.yaml
```

2. Bumped the version of the tutorial map in `triplea_maps.yaml` to `2`:

```yaml
- mapName: Tutorial
  mapCategory: BEST
  url: https://github.com/triplea-maps/tutorial/archive/master.zip
  version: 2
  description: |
    <br>This is a tutorial for players new to TripleA and its game engine.
```

3. Deleted the `TRIPLEA_LAST_CHECK_FOR_MAP_UPDATES` key from my user preferences in `~/.java/.userPrefs/games/strategy/triplea/settings/prefs.xml`:

```xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE map SYSTEM "http://java.sun.com/dtd/preferences.dtd">
<map MAP_XML_VERSION="1.0">
  <entry key="TRIPLEA_LAST_CHECK_FOR_ENGINE_UPDATE" value="2017:129"/>
  <entry key="TRIPLEA_PROMPT_TO_DOWNLOAD_TUTORIAL_MAP" value="false"/>
</map>
```

After starting TripleA, I was prompted to update my tutorial map.  (I also noted that I didn't see any `302`s, so that leads me to believe you are definitely not using your local copy of `triplea_maps.yaml`).

simon33-2 commented:
It was the third point that was missing for me. Reverting that value caused the change to work.

ssoloff commented:
And my statement about the `302`s implying you were not using your local copy was totally wrong.  During my testing, I thought the `302`s were coming from the download of `triplea_maps.yaml`, but they were actually coming from something totally different (see #1697).

RoiEXLab commented:
About fixing the 302 errors:
I'd recommend using a redirect handler similar to #1656

DanVanAtta commented on src/main/java/games/strategy/engine/framework/map/download/MapDownloadController.java:
```diff
@@ -46,16 +49,15 @@ public boolean checkDownloadedMapsAreLatest() {
+      if (!outOfDateMapNames.isEmpty()) {
+        final StringBuilder text = new StringBuilder();
+        text.append("<html>Some of the maps you have are out of date, and newer versions of those maps exist.<br>");
+        text.append("Would you like to update (re-download) the following maps now?:<br><ul>");
```
Fixing the text to match the behavior,  "would you like to open the download maps window now?" - is downright a bit funny.
That is a problem to solve, solving it now makes sense I think while we are on this topic. 

For consideration, the dialog could have the following three options:
- no/cancel/close window
- open download maps window
- update all

A simplified version of that would be to simply update the current behavior to not only open the window, but also start the download.

Until recently only one map could be downloaded at a time, and it was a UI blocking event. IN that sense it makes sense why this behavior is not a regression, the capability to start downloading multiple maps at once was not there (unless we wanted to be really annoying to the user and block the UI repeatedly)

simon33-2 commented:
I agree with Dan, although I think this step can be taken without it.

ssoloff commented:
@DanVanAtta 

> What are your thoughts on having this prompt actually initiate the update?

I looked into this briefly.  As you thought, it appears that the plumbing within `DownloadMapsWindow` already supports initiating multiple map downloads at once (see `MapDownloadProgressPanel#download(List<DownloadFileDescription>)`.  However, the `DownloadMapsWindow#showDownloadMapsWindow()` overloads only support specifying zero or one map name via an `Optional<String>`.

I think it would be relatively straightforward to modify `DownloadMapsWindow#showDownloadMapsWindow(Optional<String>)` to accept a `Collection<String>` instead, which would allow the caller to specify multiple maps to download.  Please give me 24 hours to prototype something.  I can easily see pulling the thread here touching a few classes.  So if it looks like it will cover a lot of surface area, I'll just come back and recommend we do it in a separate PR (and modify the prompt text temporarily in this PR, as you recommended above).

ssoloff commented:
@DanVanAtta Ok, that was easier than I thought...

Here's a minimal [patch](https://github.com/triplea-game/triplea/files/991904/DownloadMapsWindow.java.patch.txt) for your review.

However, there's good news and bad news.  The good news:

* Downloading multiple out-of-date maps works as expected after answering Yes to the prompt.
* If any maps cannot be downloaded (because they are not listed in `triplea_maps.yml` for some reason), a single message is displayed listing all unknown maps (while any known maps continue to be downloaded in the background).
* It looks like the only class I will have to modify is `DownloadMapsWindow`.

The bad news:

As you can see at the end of the patch, this is an incomplete fix.  `DownloadMapsWindow` will attempt to select the map (singular!) specified for download in the UI if one was provided.  Now that we may be requesting multiple maps for download, I wasn't sure if selecting multiple maps in the UI was appropriate.  Technically, selecting multiple maps should be possible since the list controls support multiple selection.  I haven't spent enough time evaluating the UX impact of such a change.  Maybe its zero...?

Anyway, I punted on the issue and simply mapped the `Collection` to an `Optional` for the remainder of the method: if the collection is empty, use an empty `Optional`; if the collection is not empty, lift the first value into an `Optional`.  Thus, only a single map will be selected even if multiple maps were specified for download.

I think selecting multiple maps is possible, but it will require significant changes to `DownloadMapsWindow`.  I've already seen one assumption that breaks down when the number of maps to select is greater than 1.  It's not insurmountable, but I don't think the changes will be as simple as those required to specify multiple maps for download in the first place. :smiley:

Thoughts?

ssoloff commented:
:+1: Good catch.  Yes, the Update tab is visible while the out-of-date maps are being downloaded:

![download-maps](https://cloud.githubusercontent.com/assets/4826349/25933974/2d385150-35e9-11e7-9738-3b241fccc1e9.png)

As you said, if we initiate the download of all out-of-date maps, the Update tab should not even be present.  I think this is simply a matter of modifying the code in `DownloadMapsWindow#createAvailableInstalledTabbedPanel()` to calculate the difference between `mapList.getOutOfDate()` and the list of maps passed to `DownloadMapsWindow#showDownloadMapsWindow(Collection)`.

ssoloff commented:
@DanVanAtta The latest commit makes the following changes:

* A new `showDownloadMapsWindow()` overload was added to `DownloadMapsWindow` that accepts a `Collection<String>` of map names to download immediately.
* When the user answers Yes to the prompt, the Download Maps window is opened and all out-of-date maps are downloaded.
* The maps that are specified for immediate download are not included on the Update tab (assuming they are considered "known" maps, i.e. they are present in `triplea_maps.yaml`).  I added the new method `MapDownloadList#getOutOfDateExcluding()` to encapsulate this logic.

Note that I punted on the multiple selection logic.  Given what you said above, it seemed relatively low value.  Please let me know if you think it should be added as part of this PR.

There was quite a bit of refactoring of `DownloadMapsWindow` in the latest commit.  Here are some highlights:

* I renamed the two overloads of `showDownloadMapsWindow()` that accept arguments (i.e. maps to download) to `showDownloadMapsWindowAndDownload()` to better describe what they are doing and to differentiate them from the no-arg version that does not initiate a download.
* I was really confused by what to name variables that represent one or more instances of `DownloadFileDescription`.  The code seems to be divided between calling them "maps", "downloads", and "games".  I settled on "downloads" simply because the factory method from which they come is named `DownloadRunnable#getDownloads()`.  So, in the methods I had to touch in this commit, I renamed existing variables with a "downloads" suffix, as needed.
* I removed the `getVersion()` method as it is no longer used.

ssoloff commented:
@DanVanAtta And I just submitted a third commit that simply renames several local variables and parameters to ensure consistency throughout the PR.

ron-murhammer commented:
@ssoloff @DanVanAtta This looks good to me. Any concerns with merging this?

ssoloff commented:
@ron-murhammer 

I think it's ready to merge, but I was waiting to see if @DanVanAtta wanted me to pursue adding multiple selection logic.  From one of my [previous comments](https://github.com/triplea-game/triplea/pull/1695#issuecomment-300973056):

> Note that I punted on the multiple selection logic. Given what you said above, it seemed relatively low value. Please let me know if you think it should be added as part of this PR.

ron-murhammer commented:
@ssoloff Given this PR is already pretty large, I'm going to move forward with merging it.

@DanVanAtta If you feel strongly about adding multiple selection logic then let's add an issue to tackle that in a future PR.
